### PR TITLE
Rewrite the interactive mode's completion engine and backtracking implementation

### LINKED ISSUE #945: Namespace info for partial completions is missing
F* is missing namespace information for some partial completions.  For example:

* `printf "module Test\nopen FStar.List\n" > test.fst`

* `fstar.exe test.fst --in`
```
#push 1 0
module Test
open FStar.List
#end #done-ok #done-nok
#completions List.append #
```

* **Output:**

```
11  List.append 
11  List.append_assoc 
11  List.append_cons_l 
11  List.append_count 
11  List.append_count_forall 
11  List.append_eq_nil 
11  List.append_eq_singl 
11  List.append_inv_head 
11  List.append_inv_tail 
11  List.append_l_cons 
11  List.append_l_nil 
11  List.append_length 
11  List.append_mem 
11  List.append_mem_forall 
11  List.append_nil_l 
11  List.append_sorted 
```

* **Expected:**

```
11 FStar List.append 
11 FStar List.append_assoc 
11 FStar List.append_cons_l 
11 FStar List.append_count 
11 FStar List.append_count_forall 
11 FStar List.append_eq_nil 
11 FStar List.append_eq_singl 
11 FStar List.append_inv_head 
11 FStar List.append_inv_tail 
11 FStar List.append_l_cons 
11 FStar List.append_l_nil 
11 FStar List.append_length 
11 FStar List.append_mem 
11 FStar List.append_mem_forall 
11 FStar List.append_nil_l 
11 FStar List.append_sorted 
```

The correct output can be seen when completing `append` instead of `list.append`:

```
6 FStar.List.Tot.Base append 
6 FStar.List.Tot.Properties append_assoc 
6 FStar.List.Tot.Properties append_cons_l 
6 FStar.List.Tot.Properties append_count 
6 FStar.List.Tot.Properties append_count_forall 
6 FStar.List.Tot.Properties append_eq_nil 
6 FStar.List.Tot.Properties append_eq_singl 
6 FStar.List.Tot.Properties append_inv_head 
6 FStar.List.Tot.Properties append_inv_tail 
6 FStar.List.Tot.Properties append_l_cons 
6 FStar.List.Tot.Properties append_l_nil 
6 FStar.List.Tot.Properties append_length 
6 FStar.List.Tot.Properties append_mem 
6 FStar.List.Tot.Properties append_mem_forall 
6 FStar.List.Tot.Properties append_nil_l 
6 FStar.List.Tot.Properties append_sorted 
```

### PR DESCRIPTION
**Pull request by cpitclaudel**
The original completion engine was incomplete (#945, #946) and ran in linear-time.  It was only able to autocomplete symbols.  It also didn't work well with module aliases.

The new implementation…

* …fixes #945 and #946 using nested tries to efficiently track lids and module names and provide substring completion (concretely, writing `module ABC = FStar.List // A.app` now properly autocompletes to `ABC.append`.
* …introduces a notion of completion context (already supported on the F*-mode side) that allows it to offer different completions in regular code, after `#set-options`, and after `open` (for that last one, it completes across all modules, not just the loaded ones, and it indicates which ones are already loaded and which ones require a reload)
* …uses that notion of context to support module names and options in `lookup` queries (used for eldoc in Emacs). 

Note that to be able to implement option completion I had to introduce a small DSL for F* options (hence the refactoring in FStar.Options).

The snapshot is included as a separate commit, which you can exclude if you're interested in looking at the diffs.

### CODE REVIEW

**Comment by cpitclaudel** [comment]
@protz I think you may be interested in https://github.com/FStarLang/FStar/pull/1260/commits/87f89730c00f82c7590c95b0632f68b79450f9ea ("ide: Clean up the backtracking implementation"), which introduces a proper stack in the interactive mode.

File: src/basic/FStar.Options.fs [hunk 1]
```diff
@@ -349,365 +352,404 @@ let mk_spec o : opt =
...
+  // For options like --extract_module that require post-processing or validation
+| Accumulated of opt_type (* elem spec *)
+  // For options like --extract_module that can be repeated (LIFO)
+| ReverseAccumulated of opt_type (* elem spec *)
+  // For options like --include that can be repeated (FIFO)
+| WithSideEffect of ((unit -> unit) * opt_type (* elem spec *))
+  // For options like --version that have side effects
+
+exception InvalidArgument of string // option name
+
```
Comment (msprotz) [inline src/basic/FStar.Options.fs]: can you comment this function and state its invariants? e.g. the "`opt_name` always matched the label of the `opt_type`, if any" and/or "this function is called upon parsing `--foo`, and receives the `opt_type` associated to `foo` as well as the argument given to the option, if any"

File: src/basic/FStar.Options.fs [hunk 2]
```diff
@@ -349,365 +352,404 @@ let mk_spec o : opt =
...
+let rec parse_opt_val (opt_name: string) (typ: opt_type) (str_val: string) : option_val =
+  try
+    match typ with
+    | Const c -> c
+    | IntStr _ -> mk_int (int_of_string str_val)
```
Comment (msprotz) [inline src/basic/FStar.Options.fs]: this will throw if the user passes an invalid integer

File: src/basic/FStar.Options.fs [hunk 3]
```diff
@@ -716,167 +758,141 @@ let rec specs () : list<Getopt.opt> =
...
         "Toggle the representation of linear arithmetic functions in the SMT encoding:\n\t\t\
          i.e., if 'boxwrap', use 'Prims.op_Addition, Prims.op_Subtraction, Prims.op_Minus'; \n\t\t\
                if 'native', use '+, -, -'; \n\t\t\
                (default 'boxwrap')");
 
        ( noshort,
         "split_cases",
-        OneArg ((fun n -> mk_int (int_of_string n)),
-                 "[positive integer]"),
-        "Partition VC of a match into groups of [n] cases");
```
Comment (msprotz) [inline src/basic/FStar.Options.fs]: is this option still used?

File: src/basic/FStar.Options.fs [hunk 4]
```diff
@@ -716,167 +758,141 @@ let rec specs () : list<Getopt.opt> =
...
-        ZeroArgs (fun () -> mk_bool true),
+        Const (mk_bool true),
         "Let the SMT solver unfold inductive types to arbitrary depths (may affect verifier performance)");
 
        ( noshort,
         "unsafe_tactic_exec",
-        ZeroArgs (fun () -> mk_bool true),
+        Const (mk_bool true),
         "Allow tactics to run external processes. WARNING: checking an untrusted F* file while \
          using this options can have disastrous effects.");
```
Comment (msprotz) [inline src/basic/FStar.Options.fs]: typo (option*s*)

File: src/tosyntax/FStar.ToSyntax.Env.fs [hunk 5]
```diff
@@ -1139,20 +1160,22 @@ let prepare_module_or_interface intf admitted env mname (mii:module_inclusion_in
     in
     let namespace_of_module = if List.length mname.ns > 0 then [ (lid_of_ids mname.ns, Open_namespace) ] else [] in
     (* [scope_mods] is a stack, so reverse the order *)
-    let auto_open = List.rev (auto_open @ namespace_of_module) in
+    let auto_open = namespace_of_module @ List.rev auto_open in
```
Comment (msprotz) [inline src/tosyntax/FStar.ToSyntax.Env.fs]: update the comment?

File: src/fstar/FStar.Interactive.Ide.fs [hunk 6]
```diff
@@ -144,26 +137,24 @@ let deps_of_our_file filename =
...
+ * returns the new environment, deps stack, and timestamps.
  *)
-let rec tc_deps (m:modul_t) (stack:stack_t)
-                (env:env_t) (remaining:list<string>) (ts:m_timestamps)
-//      : stack<'env,modul_t> * 'env * m_timestamps
+let rec tc_deps (deps_stack: deps_stack_t) (env: env_t)
+                (remaining: list<string>) (ts: m_timestamps)
   = match remaining with
-    | [] -> stack, env, ts
+    | [] -> env, (deps_stack, ts)
```
Comment (msprotz) [inline src/fstar/FStar.Interactive.Ide.fs]: can we pair the `deps_stack` with the `ts` (i.e. `List.combine` them) to enforce the invariant that they always have the same length?

File: src/basic/FStar.Options.fs (hunk 2 above)
Comment (cpitclaudel) [reply src/basic/FStar.Options.fs]: Indeed. Not sure what to do; this was like that before.

Comment (cpitclaudel) [reply src/basic/FStar.Options.fs]: I guess I could make a safe version of int_of_string

Comment (cpitclaudel) [reply src/basic/FStar.Options.fs]:
```
let safe_int_of_string x = try Some (int_of_string x) with Invalid_argument _ -> None
```

File: src/basic/FStar.Options.fs [hunk 7]
```diff
@@ -349,365 +352,404 @@ let mk_spec o : opt =
...
+  // --z3rlimit 5
+| BoolStr
+  // --admit_smt_queries true
+| PathStr of string (* label *)
+  // --fstar_home /build/fstar
+| SimpleStr of string (* label *)
+  // --admit_except xyz
+| EnumStr of list<string>
+  // --codegen OCaml
+| OpenEnumStr of list<string> (* values *) * string (* label *)
```
[the remark this answers is not in the document]
Comment (cpitclaudel) [reply src/basic/FStar.Options.fs]: Done

File: src/basic/FStar.Options.fs (hunk 3 above)
Comment (cpitclaudel) [reply src/basic/FStar.Options.fs]: No idea :)

File: src/tosyntax/FStar.ToSyntax.Env.fs (hunk 5 above)
Comment (cpitclaudel) [reply src/tosyntax/FStar.ToSyntax.Env.fs]: No, I think (?) it's correct (it's `scope_mods` that's a stack, not auto_open; but auto_open is used to populate `scope_mods` later in that function).

File: src/basic/FStar.Options.fs (hunk 1 above)
Comment (cpitclaudel) [reply src/basic/FStar.Options.fs]: Ok, I've added a docstring, with an example.

File: src/fstar/FStar.Interactive.Ide.fs (hunk 6 above)
Comment (cpitclaudel) [reply src/fstar/FStar.Interactive.Ide.fs]: Oh yes we can ^^ I take this to mean you're burning to review [cpitclaudel_912](https://github.com/FStarLang/FStar/commits/cpitclaudel_912), which does that and more ? :)

File: src/basic/FStar.Options.fs (hunk 4 above)
Comment (cpitclaudel) [reply src/basic/FStar.Options.fs]: Fixed

**Pull request merged**